libpromises/expand.c

```diff
@@ -188,13 +188,15 @@ static void MapIteratorsFromRval(EvalContext *ctx,
 }
 
 static PromiseResult ExpandPromiseAndDo(EvalContext *ctx, PromiseIterator *iterctx,
-                                        PromiseActuator *act_on_promise, void *param)
+                                        PromiseActuator *act_on_promise, void *param,
+                                        bool do_ifelse)
 {
     PromiseResult result = PROMISE_RESULT_SKIPPED;
 
     /* TODO this loop could be completely skipped for for non vars/classes if
      *      act_on_promise is CommonEvalPromise(). */
-    while (PromiseIteratorNext(iterctx, ctx))
+    bool done = ! do_ifelse;
+    while (PromiseIteratorNext(iterctx, ctx) || !done)
```

Comment:
`|| !done`? Shouldn't the loop stop when done no matter what?

Reply:
fixed